File: design-pattern-python/design_pattern_python/abstract_factory/list.py
from factory import Factory, Link, Page, Tray


class ListLink(Link):
    def __init__(self, caption: str, url: str) -> None:
        super().__init__(caption, url)

    def makeHTML(self) -> str:
        return '  <li><a href="' + self.url + '">' + self.caption + "</a></li>\n"


class ListTray(Tray):
    def __init__(self, caption) -> None:
        self.__string = ""
        super().__init__(caption)

    def makeHTML(self) -> str:
        self.__string += "<li>\n"
        self.__string += self.caption
        self.__string += "\n<ul>\n"
        for item in self.tray:
            self.__string += item.makeHTML()
        self.__string += "</ul>\n"
        self.__string += "</li>\n"
        return self.__string


class ListPage(Page):
    def __init__(self, title: str, author: str) -> None:
        self.__string = ""
        super().__init__(title, author)

    def makeHTML(self) -> str:
        self.__string += "<!DOCTYPE html>\n"
        self.__string += "<html><head><title>"
        self.__string += self.title
        self.__string += "</title></head>\n"
        self.__string += "<body>\n"
        self.__string += "<h1>\n"
        self.__string += self.title
        self.__string += "</h1>\n"
        self.__string += "<ul>\n"
        for item in self.content:
            self.__string += item.makeHTML()
        self.__string += "</ul>\n"
        self.__string += "<hr><address>"
        self.__string += self.author
        self.__string += "</address>\n"
        self.__string += "</body></html>\n"

        return self.__string


class ListFactory(Factory):
    def create_link(self, caption: str, url: str) -> Link:
        return ListLink(caption, url)

    def create_tray(self, caption: str) -> Tray:
        return ListTray(caption)

    def create_page(self, title: str, author: str) -> Page:
        return ListPage(title, author)

File: design-pattern-go/abstract_factory/list/list.go
package list

import (
	f "design-pattern-go/abstract_factory/factory"
)

/*****************************
 * ListLink
 ****************************/
type ListLink struct {
	f.Link
}

func NewListLink(caption, url string) *ListLink {
	return &ListLink{
		*f.NewLink(caption, url),
	}
}

func (l *ListLink) MakeHTML() string {
	return "  <li><a href=\"" + l.Url + "\">" + l.Item.Caption + "</a></li>\n"
}

/*****************************
 * ListTray
 ****************************/
type ListTray struct {
	f.Tray
	str string
}

func NewListTray(caption string) *ListTray {
	return &ListTray{
		Tray: *f.NewTray(caption),
		str:  "",
	}
}

func (l *ListTray) MakeHTML() string {
	l.str += "<li>\n"
	l.str += l.Caption
	l.str += "\n<ul>\n"

	for _, item := range l.Tray.Tray {
		l.str += item.MakeHTML()
	}

	l.str += "</ul>\n"
	l.str += "</li>\n"

	return l.str
}

/*****************************
 * ListPage
 ****************************/
type ListPage struct {
	f.Page
	str string
}

func NewListPage(title, author string) *ListPage {
	return &ListPage{
		Page: *f.NewPage(title, author),
		str:  "",
	}
}

func (l *ListPage) MakeHTML() string {
	l.str += "<!DOCTYPE html>\n"
	l.str += "<html><head><title>"
	l.str += l.Title
	l.str += "</title></head>\n"
	l.str += "<body>\n"
	l.str += "<h1>\n"
	l.str += l.Title
	l.str += "</h1>\n"
	l.str += "<ul>\n"

	for _, item := range l.Content {
		l.str += item.MakeHTML()
	}

	l.str += "</ul>\n"
	l.str += "<hr><address>"
	l.str += l.Author
	l.str += "</address>"
	l.str += "</body></html>\n"

	return l.str
}

/*****************************
 * ListFactory
 ****************************/
type ListFactory struct{}

func NewListFactory() f.FactoryInterface {
	return &ListFactory{}
}

func (l *ListFactory) CreateLink(caption, url string) f.LinkInterface {
	return NewListLink(caption, url)
}

func (l *ListFactory) CreateTray(caption string) f.TrayInterface {
	return NewListTray(caption)
}

func (l *ListFactory) CreatePage(title, author string) f.PageInterface {
	return NewListPage(title, author)
}
